enumerables.rb

```diff
@@ -0,0 +1,155 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+
+module Enumerable
+  # Task-1 my_each exercise
+  def my_each
+    if block_given?
+      length.times { |e| yield(self[e]) }
+    else
+      to_enumerator(:my_each)
+    end
+  end
+
+  # Test for my_each
+  # [1,2,3].my_each {|e| p "#{e}"}
+  # Task-2 my_each_with_index
+  def my_each_with_index
+    if block_given?
+      (0...length).each { |index| yield(self[index], index) }
+    else
+      to_enumerator(:my_each_with_index)
+    end
+  end
+
+  # Test for my_each
+  # [1,2,3].my_each_with_index {|e,i| p "#{e}:#{i}"}
+  # Task-3 my_select
+  def my_select
+    if block_given?
+      arr = []
+      (0...length).each { |element| arr.push(element) if yield(element) }
+      arr
+    else
+      to_enum(:my_select)
+    end
+  end
+
+  # Test for my_select
+  # p([1, 2, 3, 4, 5, 6, 7, 8].my_select { |x| x.even? })
+  # Task 4 my_all
+  def my_all
+    if block_given?
+      result = []
+      (0...length).each { |element| result.push(yield(self[element])) }
+      p false if result.include?(false)
+      p true if result.include?(true)
+    else
+      to_enumerator(:my_all)
+    end
+  end
```

Comment:
#my_all? returns true if the block never returns false or nil
#my_all? returns false if the block returns false or nil
#my_all? returns false if the block returns false or nil::range
#my_all? does not mutate the original array
#my_all? when no block or argument is given returns true when none of the collection members are false or nil
#my_all? when no block or argument is given returns false when one of the collection members are false or nil
#my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
#my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
#my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
#my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
#my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
#my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
#my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern